File: Android /Android/Separate Android /Android/JMBA_Olo_Android_SDK/app/src/main/java/com/fishbowl/apps/olo/Services/OloUserService.java
package com.fishbowl.apps.olo.Services;

import com.google.gson.Gson;
import com.fishbowl.apps.olo.Interfaces.OloBillingAccountCallback;
import com.fishbowl.apps.olo.Interfaces.OloServiceCallback;
import com.fishbowl.apps.olo.Interfaces.OloUserContactDetailCallback;
import com.fishbowl.apps.olo.Interfaces.OloUserOrderServiceCallback;
import com.fishbowl.apps.olo.Interfaces.OloUserPostServiceCallback;
import com.fishbowl.apps.olo.Interfaces.OloUserServiceCallback;
import com.fishbowl.apps.olo.Models.OloBillingAccount;
import com.fishbowl.apps.olo.Models.OloOrderStatus;
import com.fishbowl.apps.olo.Models.OloUser;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;

import static com.fishbowl.apps.olo.Utils.Constants.FORGOT_PASSWORD;

/**
 * Created by Nauman Afzaal on 27/04/15.
 */
public class OloUserService
{
    public static void changePassword(String currentPassword, String newPassword, final OloUserPostServiceCallback callback)
    {
        String authToken = OloSessionService.currentSesstion.getAuthToken();
        if (authToken == null)
        {
            callback.onUserPostCallback(false, new Exception("User is not authenticated."));
            return;
        }
        String path = "users/" + authToken + "/password";
        HashMap<String, Object> parameters = new HashMap<>();
        parameters.put("currentpassword", currentPassword);
        parameters.put("newpassword", newPassword);

        OloService.getInstance().post(path, parameters, new OloServiceCallback()
        {
            @Override
            public void onServiceCallback(JSONObject response, Exception error)
            {
                boolean isPasswordChanged = error == null;
                if (callback != null)
                {
                    callback.onUserPostCallback(isPasswordChanged, error);
                }
            }
        });
    }

    public static void forgotPassword(String email, final OloUserPostServiceCallback callback)
    {
        HashMap<String, Object> parameters = new HashMap<>();
        parameters.put("emailaddress", email);

        OloService.getInstance().post(FORGOT_PASSWORD, parameters, new OloServiceCallback()
        {
            @Override
            public void onServiceCallback(JSONObject response, Exception error)
            {
                boolean isPasswordChanged = error == null;
                if (callback != null)
                {
                    callback.onUserPostCallback(isPasswordChanged, error);
                }
            }
        });
    }

    public static void getUserInformation(final OloUserServiceCallback callback)
    {
        String authToken = OloSessionService.currentSesstion.getAuthToken();
        if (authToken == null)
        {
            callback.onUserServiceCallback(null, new Exception("User is not authenticated."));
            return;
        }
        String path = "users/" + authToken;
        OloService.getInstance().get(path, null, new OloServiceCallback() {
            @Override
            public void onServiceCallback(JSONObject response, Exception error) {
                if (callback != null) {
                    parseResponse(response, error, callback);
                }
            }
        });
    }

    public static void updateUserInformation(OloUser user, final OloUserServiceCallback callback)
    {
        String authToken = OloSessionService.currentSesstion.getAuthToken();
        if (authToken == null)
        {
            callback.onUserServiceCallback(null, new Exception("User is not authenticated."));
            return;
        }
        String path = "users/" + authToken;
        HashMap<String, Object> parameters = new HashMap<>();
        parameters.put("firstname", user.getFirstname());
        parameters.put("lastname", user.getLastname());
        parameters.put("cardsuffix", user.getCardsuffix());
        parameters.put("emailaddress", user.getEmailaddress());

        OloService.getInstance().put(path, parameters, new OloServiceCallback() {
            @Override
            public void onServiceCallback(JSONObject response, Exception error) {
                if (callback != null) {
                    parseResponse(response, error, callback);
                }
            }
        });
    }

    public static void addCreditCard(String cardNumber, int expiryMonth, int expiryYear, String cvv, String zip, final OloUserPostServiceCallback callback)
    {
        String authToken = OloSessionService.currentSesstion.getAuthToken();
        if (authToken == null)
        {
            callback.onUserPostCallback(false, new Exception("User is not authenticated."));
            return;
        }
        String path = "users/" + authToken + "/creditcard";
        HashMap<String, Object> parameters = new HashMap<>();
        parameters.put("cardnumber", cardNumber);
        parameters.put("expirymonth", expiryMonth);
        parameters.put("expiryyear", expiryYear);
        parameters.put("cvv", cvv);
        parameters.put("zip", zip);
        OloService.getInstance().post(path, parameters, new OloServiceCallback() {
            @Override
            public void onServiceCallback(JSONObject response, Exception error) {
                boolean isSucces = error == null;
                if (callback != null) {
                    callback.onUserPostCallback(isSucces, error);
                }
            }
        });
    }

    public static void deleteCreditCard(String cardNumber, final OloUserPostServiceCallback callback)
    {
        String authToken = OloSessionService.currentSesstion.getAuthToken();
        if (authToken == null)
        {
            callback.onUserPostCallback(false, new Exception("User is not authenticated."));
            return;
        }
        String path = "users/" + authToken + "/creditcard";
        HashMap<String, Object> parameters = new HashMap<>();
        parameters.put("path", cardNumber);
        OloService.getInstance().delete(path, parameters, new OloServiceCallback() {
            @Override
            public void onServiceCallback(JSONObject response, Exception error) {
                boolean isSucces = error == null;
                if (callback != null) {
                    callback.onUserPostCallback(isSucces, error);
                }
            }
        });
    }

    public static void getRecentOrders(final OloUserOrderServiceCallback callback)
    {
        String authToken = OloSessionService.currentSesstion.getAuthToken();
        if (authToken == null)
        {
            callback.onUserOrderServiceCallback(null, new Exception("User is not authenticated."));
            return;
        }
        String path = "users/" + authToken + "/recentorders";
        OloService.getInstance().get(path, null, new OloServiceCallback() {
            @Override
            public void onServiceCallback(JSONObject response, Exception error) {
                OloOrderStatus[] orderStatuses = null;
                if (response != null) {
                    Gson gson = new Gson();
                    try {
                        JSONArray jsonArray = response.getJSONArray("orders");
                        orderStatuses = gson.fromJson(jsonArray.toString(), OloOrderStatus[].class);
                    } catch (JSONException e) {
                        e.printStackTrace();
                        error = new Exception("Error occurred while parsing data.");
                    }
                }
                if (callback != null) {
                    callback.onUserOrderServiceCallback(orderStatuses, error);
                }
            }
        });
    }

    public static void getUserContactDetail(final OloUserContactDetailCallback callback)
    {
        String authToken = OloSessionService.currentSesstion.getAuthToken();
        if (authToken == null)
        {
            callback.onUserContactDetailCallback(null, new Exception("User is not authenticated."));
            return;
        }

        String path = "users/" + authToken + "/contactdetails";
        OloService.getInstance().get(path, null, new OloServiceCallback() {
            @Override
            public void onServiceCallback(JSONObject response, Exception error) {
                String contactdetails = null;
                if (response != null) {
                    try {
                        contactdetails = response.getString("contactdetails");

                    } catch (JSONException e) {
                        e.printStackTrace();
                        error = new Exception("Error occurred while parsing data.");
                    }
                }
                if (callback != null) {
                    callback.onUserContactDetailCallback(contactdetails, error);
                }
            }
        });
    }

    public static void updateUserContactDetail(String contactDetail, final OloUserContactDetailCallback callback)
    {
        String authToken = OloSessionService.currentSesstion.getAuthToken();
        if (authToken == null)
        {
            callback.onUserContactDetailCallback(null, new Exception("User is not authenticated."));
            return;
        }
        String path = "users/" + authToken + "/contactdetails";
        HashMap<String, Object> parameters = new HashMap<>();
        parameters.put("contactdetails", contactDetail);
        OloService.getInstance().put(path, parameters, new OloServiceCallback() {
            @Override
            public void onServiceCallback(JSONObject response, Exception error) {
                String contactdetails = null;
                if (response != null) {
                    try {
                        contactdetails = response.getString("contactdetails");

                    } catch (JSONException e) {
                        e.printStackTrace();
                        error = new Exception("Error occurred while parsing data.");
                    }
                }
                if (callback != null) {
                    callback.onUserContactDetailCallback(contactdetails, error);
                }
            }
        });
    }

    private static void parseResponse(JSONObject response, Exception error, OloUserServiceCallback callback)
    {
        OloUser user = null;
        if (response != null)
        {
            Gson gson = new Gson();
            if (response.has("user"))
            {
                try
                {
                    response = response.getJSONObject("user");
                } catch (JSONException e)
                {
                    e.printStackTrace();
                }
            }
            user = gson.fromJson(response.toString(), OloUser.class);
            if (user == null)
            {
                error = new Exception("Error occurred while parsing data.");
            }
        }
        if (callback != null)
        {
            callback.onUserServiceCallback(user, error);
        }
    }

    public static void getBillingAccount(final OloBillingAccountCallback callback)
    {
        String authToken = OloSessionService.currentSesstion.getAuthToken();
        if(authToken == null)
        {
            callback.onBillingAccountCallback(null, new Exception("User is not authenticated."));
            return;
        }
        String path = "users/" + authToken + "/billingaccounts";
        OloService.getInstance().get(path, null, new OloServiceCallback()
        {
            @Override
            public void onServiceCallback(JSONObject response, Exception error)
            {
                OloBillingAccount[] accounts = null;
                if (response != null)
                {
                    Gson gson = new Gson();
                    try
                    {
                        JSONArray jsonArray = response.getJSONArray("userbillingaccounts");
                        accounts = gson.fromJson(jsonArray.toString(), OloBillingAccount[].class);
                    } catch (JSONException e)
                    {
                        e.printStackTrace();
                        error = new Exception("Error occurred while parsing data.");
                    }
                }
                if (callback != null)
                {
                    callback.onBillingAccountCallback(accounts, error);
                }
            }
        });
    }
}

File: Android /Android/25:Jan:2017/FishbowlMobileApps/AndroidSDK/FishbowlMobileApps-Jamba-Android-v2.4.1/JMBA_Olo_iOS_SDK/Olo/OloUserService.swift
//
//  OloUserService.swift
//  JambaJuice
//
//  Created by Eneko Alonso on 4/16/15.
//  Copyright (c) 2015 Jamba Juice. All rights reserved.
//

import UIKit
import SwiftyJSON

public typealias OloUserCallback = (_ user: OloUser?, _ error: NSError?) -> Void
public typealias OloOrderStatusCallback = (_ status: [OloOrderStatus], _ error: NSError?) -> Void
public typealias OloUserContactDetailsCallback = (_ details: String?, _ error: NSError?) -> Void


open class OloUserService {
   
    open class func forgotPassword(_ emailAddress: String, callback: @escaping OloErrorCallback) {
        let payload = [
            "emailaddress": emailAddress
        ]
        OloService.post("/users/forgotpassword", parameters: payload) { (response, error) -> Void in
            callback(error)
        }
    }
    
    open class func changePassword(_ currentPassword: String, newPassword: String,callback: @escaping OloErrorCallback) {
        let authToken = OloSessionService.authToken
        if authToken == nil {
            callback(OloUtils.error("User is not authenticated.", code: 0))
            return
        }

        let payload = [
            "currentpassword": currentPassword,
            "newpassword": newPassword
        ]
        OloService.post("/users/\(authToken!)/password", parameters: payload) { (response, error) -> Void in
            callback(error)
        }
    }
    
    open class func userInformation(_ callback: @escaping OloUserCallback) {
        let authToken = OloSessionService.authToken
        if authToken == nil {
            callback(nil, OloUtils.error("User is not authenticated.", code: 0))
            return
        }

        OloService.get("/users/\(authToken!)", parameters: nil) { (response, error) -> Void in
            if error != nil {
                callback(nil, error)
                return
            }
            let user = OloUser(json: response)
            OloSessionService.authToken = user.authToken
            callback(user, nil)
        }
    }
    
    open class func updateUserInformation(_ user: OloUser, callback: @escaping OloUserCallback) {
        let authToken = OloSessionService.authToken
        if authToken == nil {
            callback(nil, OloUtils.error("User is not authenticated.", code: 0))
            return
        }

        var payload:[String:AnyObject] = [
            "firstname": user.firstName as AnyObject,
            "lastname": user.lastName as AnyObject,
            "emailaddress": user.emailAddress as AnyObject
        ]
        if user.cardSuffix != nil {
            payload["cardsuffix"] = user.cardSuffix! as AnyObject?
        }
        else {
            payload["cardsuffix"] = NSNull()
        }
        OloService.put("/users/\(authToken!)", parameters: payload) { (response, error) -> Void in
            if error != nil {
                callback(nil, error)
                return
            }
            let user = OloUser(json: response)
            OloSessionService.authToken = user.authToken
            callback(user, nil)
        }
    }
    
    open class func recentOrders(_ callback: @escaping OloOrderStatusCallback) {
        let authToken = OloSessionService.authToken
        if authToken == nil {
            callback([], OloUtils.error("User is not authenticated.", code: 0))
            return
        }

        OloService.get("/users/\(authToken!)/recentorders", parameters: nil) { (response, error) -> Void in
            if error != nil {
                callback([], error)
                return
            }
            let orderStatusArray = response["orders"].arrayValue.map { item in OloOrderStatus(json: item) }
            callback(orderStatusArray, nil)
        }
    }
    
    open class func addCreditCard(_ cardNumber: String, expiryMonth: Int, expiryYear: Int, cvv: String, zip: String, callback: @escaping OloErrorCallback) {
        let authToken = OloSessionService.authToken
        if authToken == nil {
            callback(OloUtils.error("User is not authenticated.", code: 0))
            return
        }

        let payload:[String:AnyObject] = [
            "cardnumber": cardNumber as AnyObject,
            "expirymonth": expiryMonth as AnyObject,
            "expiryyear": expiryYear as AnyObject,
            "cvv": cvv as AnyObject,
            "zip": zip as AnyObject
        ]
        OloService.post("/users/\(authToken!)/creditcard", parameters: payload) { (response, error) -> Void in
            callback(error)
        }
    }
  
    open class func deleteCreditCard(_ callback: @escaping OloErrorCallback) {
        let authToken = OloSessionService.authToken
        if authToken == nil {
            callback(OloUtils.error("User is not authenticated.", code: 0))
            return
        }
        OloService.delete("/users/\(authToken!)/creditcard", parameters: nil) { (response, error) -> Void in
            callback(error)
        }
    }

    open class func contactDetails(_ callback: @escaping OloUserContactDetailsCallback) {
        let authToken = OloSessionService.authToken
        if authToken == nil {
            callback(nil, OloUtils.error("User is not authenticated.", code: 0))
            return
        }
        OloService.get("/users/\(authToken!)/contactdetails", parameters: nil) { (response, error) -> Void in
            if error != nil {
                callback(nil, error)
                return
            }
            else {
                let contactDetails = response["contactdetails"].stringValue
                callback(contactDetails, nil)
            }
        }
    }
    
    open class func updateContactDetails(_ contactDetails: String, callback: @escaping OloUserContactDetailsCallback) {
        let authToken = OloSessionService.authToken
        if authToken == nil {
            callback(nil, OloUtils.error("User is not authenticated.", code: 0))
            return
        }
        let payload:[String:AnyObject] = [
            "contactdetails": contactDetails as AnyObject
        ]
        OloService.put("/users/\(authToken!)/contactdetails", parameters: payload) { (response, error) -> Void in
            if error != nil {
                callback(nil, error)
                return
            }
            else {
                let contactDetails = response["contactdetails"].stringValue
                callback(contactDetails, nil)
            }
        }
    }
    
    //MARK:- Dispatch Address
    open class func getSavedDeliveryaddressList(_ callback: @escaping OloBasketDeliveryAddressListCallback) {
        let authToken = OloSessionService.authToken
        guard let token = authToken else {
            callback([], OloUtils.error("User is not authenticated.", code: 0))
            return
        }
        
        OloService.get("/users/\(token)/userdeliveryaddresses") { (response, error) -> Void in
            if error != nil {
                callback([], error)
                return
            }
            let address = OloUserDeliveryAddressList(json: response)
            callback(address.deliverAddress, nil)
        }
    }
    
    
    open class func deleteSavedDeliveryaddressList(_ addressId:String,callback: @escaping OloErrorCallback) {
        let authToken = OloSessionService.authToken
        guard let token = authToken else {
            callback(OloUtils.error("User is not authenticated.", code: 0))
            return
        }
        
        OloService.delete("/users/\(token)/userdeliveryaddresses/\(addressId)") { (response, error) -> Void in
            if error != nil {
                callback(error)
                return
            }
            callback(nil)
        }
    }
}

